[ONNX] Support custom Op with onnx-script local function

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86906

Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .

titaiwangms committed:
[ONNX] Support custom Op with onnx-script local function

[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/utils.py:
```diff
@@ -1663,6 +1667,15 @@ def _export(
+def _add_onnxscript_fn(model_proto: bytes) -> bytes:
+    model = onnx.load_model_from_string(model_proto)
+    if _custom_onnxscript is not None:
+        model.functions.extend(_custom_onnxscript)
```
Wondering how I can do this in Cpp. In that case, we don't need onnx for dependency ...

BowenBao replied:
checkout https://github.com/pytorch/pytorch/blob/c43033edec898aa5e470d955873d9b175d272be9/torch/csrc/jit/serialization/export.cpp#L776
and
https://github.com/pytorch/pytorch/blob/c43033edec898aa5e470d955873d9b175d272be9/torch/csrc/jit/python/python_ir.cpp#247
If you want you can implement this in C++ and expose through pybind.

BowenBao replied:
While I'm curious, does onnx-script depend on onnx? and do we want to add onnx-script dependency to pytorch?

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1663,6 +1667,15 @@ def _export(
+def _add_onnxscript_fn(model_proto: bytes) -> bytes:
+    model = onnx.load_model_from_string(model_proto)
+    if _custom_onnxscript is not None:
+        model.functions.extend(_custom_onnxscript)
```
Discussed offline, we should update the logic to include only the function proto used in the model.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -72,6 +72,7 @@ def is_in_onnx_export() -> bool:
 # TODO(justinchuby): Remove dependency to this global variable from constant_fold.cpp
 # Skip check due to cannot import IValue from torch._C
 _params_dict = {}  # type: ignore[var-annotated]
+_custom_onnxscript = list()  # type: ignore[var-annotated]
```
I hope we can find a cleaner way to store the state than using a global var. Do we have options?

AllenTiTaiWang replied:
Yes, I will reference on registry and see if we can have one data structure for onnx-script function.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -11,8 +11,15 @@
+try:
+    import onnxscript  # type: ignore[import]
+except Exception:
+    raise errors.OnnxExporterError("Module onnxscript is not installed!")
```
We should prob just do nothing if onnxscript is not found

AllenTiTaiWang replied:
I removed the whole thing, since it was only for annotation.

justinchuby commented on torch/onnx/symbolic_helper.py:
```diff
@@ -439,6 +439,12 @@ def _if_scalar_type_as(self, tensor):
     return self
 
 
+@_beartype.beartype
+def _is_custom_domain(node: _C.Node) -> bool:
+    domain = node.kind().split("::")[0]
+    return domain not in {"onnx", "aten", "prim"}
```
We should move away from using private methods outside of their own modules. In this case we can add it to jit_utils. I would also consider decoupling string stuff from Nodes. I will take a closer look later. Happy to discuss

AllenTiTaiWang replied:
Agree and will move it. I actually thought about adding this into jit_utils.py, but then I saw lots of node processing and bool function in symbolic_helper.py, so I went with it.

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1598,7 +1597,7 @@ def _export(
                     node_attr_to_name,
                 )
-            proto = _add_onnxscript_fn(proto)
+            proto = _add_onnxscript_fn(proto, graph, custom_opsets)
```
If a user exports two different models that use different functions, would this add all functions to the second model?

AllenTiTaiWang replied:
I don't think so. There is a logic inside _add_onnxscript_fn making sure only the model-related function is inserted.

AllenTiTaiWang replied:
Although a complicated test case may be added to make sure this won't happen.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1661,50 @@ def _export(
+    # TODO(titaiwang): remove this when onnx becomes dependency
+    try:
+        import onnx
+    except Exception:
```
```suggestion
    except ImportError:
```

I would avoid catching too generic errors, as errors like MemoryError, KeyboardInterrupt etc. should still need to manifest as themselves.

AllenTiTaiWang replied:
Done

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1661,50 @@ def _export(
+    # function_proto into model_proto
+    for node in graph.nodes():
+        if symbolic_helper._is_custom_domain(node):
+            domain_op = node.kind()
```
If we are calling node.kind() anyways, can is_custom_domain operate on a string instead?

AllenTiTaiWang replied:
Done

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1661,50 @@ def _export(
+    for node in graph.nodes():
+        if symbolic_helper._is_custom_domain(node):
+            domain_op = node.kind()
+            domain, opname = domain_op.split("::")
```
I am seeing more places where we need to split the string. I would consider creating a util somewhere (symbolic_helpers?)

If we decided symbolic_helpers is the place, it needs to be public as symbolic_helpers itself is public and we don't want to use private functions in a module where it is not defined.

AllenTiTaiWang replied:
#87638 
Personally I feel like this function is too easy that no external users would care. I suugest if there is function that we are not sure about going public, we just make it internal first. What do you think?

justinchuby replied:
SG. Commented in the PR!

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1661,50 @@ def _export(
+                if onnx_fn is not None:
+                    # TODO: to_function_proto is onnx-script API and can be annotated
+                    # after onnx-script is dependency
+                    onnx_function_list.append(onnx_fn.to_function_proto())  # type: ignore[attr-defined]
```
Do we need to deduplicate this list?

justinchuby replied:
What is the current behavior if the same `onnx_fn` is seen twice? Do we append two of them in the list?

AllenTiTaiWang replied:
Ahh I see

AllenTiTaiWang replied:
Done. I used a set. LMK if there is a better approach on your mind.

gramalingam replied:
Since this is doing a sort of linking-step, it would be good to pin down the desired behavior: if an added function calls other onnx functions (transitively), should they be linked as well or not?

AllenTiTaiWang replied:
IIRC, @BowenBao brought this up in the meeting, and I remember we don't support this yet. So no, this doesn't include function calling function behavior. I will comment this on this logic.

AllenTiTaiWang commented on torch/onnx/utils.py:
```diff
@@ -1663,6 +1667,15 @@ def _export(
 
+@_beartype.beartype
+def _add_onnxscript_fn(model_proto: bytes) -> bytes:
+    model = onnx.load_model_from_string(model_proto)
```
It doesn't. I will have to figure out how to get >2GB models working.

AllenTiTaiWang replied:
@BowenBao I did some dig out in >2GB scenario. I have some questions.

1. I found that export.cpp actually handles >2GB situation. It seems to save tensor proto in the same directory with onnx file, but in separate external data files, which is an ONNX suggesting way based on ONNX repo. And I figure in this circumstance, the return proto should be <2GB by then, as we already separately save the tensors from model. Otherwise, it's no benefit of doing that.
2. I think ONNX repo doesn't test 2GB case, and neither do we. I think it's because of CI? Do I need to add test case for 2GB after I implement this?

AllenTiTaiWang replied:
I guess this is hard to verify anyway. I ended up adding a block for >2GB model to load without tensor files just in case my assumption that **model bytes is impossible > 2GB in this point** is inaccurate. LMK what you think.

AllenTiTaiWang replied:
Tested with make up model > 2GB with @justinchuby , and the behavior is super weird in `torch.onnx.export` lol

AllenTiTaiWang replied:
The detail is in description.

xadupre commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -98,6 +99,49 @@ def aten_op(self, operator: str, *args, overload_name: str = "", **kwargs):
+        self,
+        onnx_fn,  # TODO(titaiwang): annotate this when onnx-script becomes dependency
+        *raw_args: Union[torch.Tensor, _C.Value],
+        outputs: int = 1,
```
I wonder if we could None here and infer that value from the annotation of the function and only specify it when it is unknown (like Split operator).

AllenTiTaiWang replied:
I tried with ast and inspect package, just like onnx-script does, but it turns out at this point the function is no longer python-function, and it's ONNXfunction instead, which is not understandable by inspect. Digging more into this seems to be complicated for this goal... Do you have any sugestion on this?

xadupre replied:
Attribute function should give ou access the the python function: https://github.com/microsoft/onnx-script/blob/main/onnxscript/values.py#L158.

gramalingam replied:
On the other hand, the outputs is passed through directly to `_add_op` and this seems to be following the same pattern as the other `op` methods above. So, maybe it doesn't matter? I wouldn't suggest analyzing the python-source-ast, it is best to leave that to ONNXScript (and just expose methods on ONNXFunction to return it's type-annotation (if available)).

AllenTiTaiWang replied:
Yes, it makes sense to keep this as the same format as `op` API. I think I will leave this one as it is.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

gramalingam commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -98,6 +99,49 @@ def aten_op(self, operator: str, *args, overload_name: str = "", **kwargs):
+        outputs: int = 1,
+        **kwargs,
+    ):
+        """Creates an ONNX operator from onnx-script function, taking "raw_args" as inputs and "kwargs" as attributes.
```
I am unclear what is the intended usage here? I thought this was for static-graph building? Can it also be used in a dynamic setting? It's probably just my lack of familiarity with GraphContext and its usage. I was wondering about the use of kwargs/args, but I guess this is just inheriting/reusing the usage-pattern of the existing op/_add_op methods, so maybe it doesn't matter.

gramalingam commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1669,64 @@ def _export(
+@_beartype.beartype
+def _add_onnxscript_fn(
+    model_bytes: bytes,
+    graph: _C.Graph,
```
Why do we need the graph as an extra input? Since the onnx model is available, can we not find out which functions are used from the onnx model itself? It seems to me that would help decouple this utility (making it independent how the model is created).

AllenTiTaiWang replied:
Great idea! We should totally use more onnx API.

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1669,10 +1627,54 @@ def _export(
 
 
+@_beartype.beartype
+def _export_file(
```
Awesome refactoring

justinchuby commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -381,4 +381,4 @@ def parse_node_kind(kind: str) -> Tuple[str, str]:
 def is_custom_domain(domain: str) -> bool:
     """Check if the domain is official."""
-    return domain not in {"onnx", "aten", "prim"}
+    return domain not in {"onnx", "aten", "prim", ""}
```
Does `""` happen?

AllenTiTaiWang replied:
Yes, cause I use onnx model proto instead, and there might be identity without domain.

justinchuby replied:
Would it be possible to normalize the name before calling this function? I think it would make the code more robust to have "always have a domain" as an invariant

AllenTiTaiWang replied:
I guess I can check if the node proto has domain field or not, but another question is that is "" one of the domain? I remember we accept that in some cases, and take them as aten/onnx for default.

AllenTiTaiWang replied:
Open a new PR to add onnx_proto_utils.py for internal utils function in utils.py

AllenTiTaiWang replied:
#88376

AllenTiTaiWang replied:
Just found another issue with this approach. The domain startswith `org.pytorch` is also official... kind of lost in the pattern we have in here...

xadupre commented on .jenkins/caffe2/test.sh:
```diff
@@ -176,7 +176,7 @@ fi
+  pip install -q --user ninja flatbuffers==2.0 numpy==1.21.5 onnxruntime==1.12.1 beartype==0.10.4 onnx==1.12.0
   # numba requires numpy <= 1.20, onnxruntime requires numpy >= 1.21.
   # We don't actually need it for our tests, but it's imported if it's present, so uninstall.
   pip uninstall -q --yes numba
```
That does not seem necessary anymore.

AllenTiTaiWang replied:
I think it still needs it cause https://github.com/pytorch/pytorch/blob/954b7ac73b7251e6c62050e5bc3869ff05b1fe95/torch/onnx/utils.py#L1685-L1694

xadupre replied:
I was thinking about the comment about numba. But it is unrelated to your changes.

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1627,115 @@ def _export(
+    # ONNXFunction calling other ONNXFunction scenario, neither does it here
+    onnx_function_list = list()
+    included_node_func = set()
+    for node in model_proto.graph.node:
```
Need to recursively check subgraph in nodes ... for operators like `onnx::Loop`, or `onnx::If` that has attribute of type `GraphProto`.

AllenTiTaiWang replied:
Ah....

AllenTiTaiWang replied:
Added a recursive function to get graphs from nodes, and a specific loop test in the next PR. LMK if you know what else we can do to make it even robust.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1630,11 +1628,14 @@ def _export(
     """export/write model bytes into directory/protobuf/zip"""
+    # TODO(titaiwang) MYPY asks for os.PathLike[str] type for parameter: f,
+    # but beartype raises beartype.roar.BeartypeDecorHintNonpepException,
+    # as os.PathLike[str] uncheckable at runtime
```
Just do `os.PathLike`?

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1653,17 +1654,17 @@ def _export_file(
     elif export_type == _exporter_states.ExportTypes.DIRECTORY:
-        if os.path.exists(f):
-            assert os.path.isdir(f)
+        if os.path.exists(f):  # type: ignore[arg-type]
```
I think you can assert f is not `io.BytesIO`

AllenTiTaiWang replied:
MYPY is not happy PathLike without specifying what's inside.

AllenTiTaiWang replied:
Like

```python
    elif export_type == _exporter_states.ExportTypes.DIRECTORY:
        assert not isinstance(io.BytesIO, f)
        assert os.path.isdir(f)  # type: ignore[arg-type]
        if not os.path.exists(f):  # type: ignore[arg-type]
            os.makedirs(f)  # type: ignore[arg-type]
```
?

AllenTiTaiWang replied:
I use raise instead, because I think these are from user? It doesn't make MYPY happy though. Every line is using os functions, it's going to complain about PathLike type. I would leave this as it is.

justinchuby replied:
hmm OK. I thought os.path.exists could take str as well. nvm

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,110 @@ def _export(
+            specified_version = custom_opsets.get(node.domain, 1)
+            onnx_function_group = registration.registry.get_function_group(node_kind)
+            if onnx_function_group is not None:
+                onnx_fn = onnx_function_group.get(specified_version)
```
One type of likely user error is the version of `onnx_fn` they passed in `onnxscript_op` doesn't match the `specified_version`.

BowenBao replied:
And how to ensure the onnx version in `onnx_fn` is compatible with that in `model_proto`?

AllenTiTaiWang replied:
So we have two opset versions here. One is custom op version, the other is opset_version of model_proto.
Let's say we register a `Selu` with custom opset version 1, and implemented with opset 15. In this case, If we call `torch.onnx.export` with opset_version < 15, this custom op would not be consumed in the model_proto, based on the current registry design and my experiment. In another case, if user wrongly call `torch.onnx.export` with custom op version 2, then he/she would fallback to custom opset version 1, based on the logic here.
LMK if I anwser this question?

BowenBao replied:
Thanks, that makes sense. I looked into #86907 to understand 
> If we call torch.onnx.export with opset_version < 15, this custom op would not be consumed in the model_proto

BowenBao commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -98,6 +99,49 @@ def aten_op(self, operator: str, *args, overload_name: str = "", **kwargs):
+        """
+        # TODO(titaiwang): This is using class attributes, and it needs to be updated
+        # if onnx-script makes any change on these.
+        symbolic_name = f"{onnx_fn.opset}::{onnx_fn.opname}"
```
Does `onnx_fn.opset` repr shows only domain or does it include version too?

gramalingam replied:
I think it would be better to explicitly encode what we want here (instead of making assumptions about `opset` repr (which might change over time).

AllenTiTaiWang replied:
@BowenBao It should be only domain, based on the passed tests in the next PR.
@gramalingam Do you mean hard encode here? If that's the case, the user wouldn't be able to use multiple domains in this kind of usage.

BowenBao replied:
I think he was referring to something like `{onnx_fn.opset.domain}`

AllenTiTaiWang replied:
I can turn in a PR in onnx-script to expose this information.

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: Did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -332,3 +376,11 @@ def parse_node_kind(kind: str) -> Tuple[str, str]:
+def is_custom_domain(domain: str) -> bool:
+    """Check if the domain is official."""
+    return domain not in {"onnx", "aten", "prim"} and not domain.startswith(
+        "org.pytorch"  # to include ATen
```
@justinchuby @BowenBao this function is not robust, as I found that there could by `org.pytorch` family staying in the onnx graph, and they look like official domain somehow. Not sure if I miss other "family here". Could you guys confirm if this logic is sufficient?

justinchuby replied:
I would separate domains in torchscript and onnx in the implementation. Is `org.pytorch` a torch concept or a onnx concept? If it is a onnx concept, I would handle it in a different place

AllenTiTaiWang replied:
TODO: move it to onnx_utils.py

AllenTiTaiWang replied:
Suggest to change

AllenTiTaiWang replied:
Removed

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

AllenTiTaiWang commented:
@BowenBao @justinchuby PTAL

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,135 @@ def _export(
+
+
+@_beartype.beartype
+def _find_onnxscript_op_recursively(
```
Add a test for this function?

justinchuby replied:
It'd be helpful to have a unit test for this `_find_onnxscript_op_recursively` function I think

justinchuby replied:
Aka calling it directly. It seems its logic is worth tested more directly.

AllenTiTaiWang replied:
Create another one in `test_utility_func.py`?

justinchuby replied:
Sounds good!

AllenTiTaiWang replied:
I found this is not easy to call alone in terms of the inputs it needs. The function is highly relying on model, and if I made up a model for it, it would be the test case I pasted above. LMK if you have a good idea on this. I think the existing one makes a fine test.

justinchuby replied:
I see. In that case would it be possible to add a few comments on this existing test to underscore that it needs to create a nested graph?

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,135 @@ def _export(
+
+
+@_beartype.beartype
+def _find_onnxscript_op_recursively(
```
Done

gramalingam commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,137 @@ def _export(
+                (
+                    onnx_function_list,
+                    included_node_func,
+                ) = _find_onnxscript_op_recursive(
```
Nit: the function destructively updates the input parameters (set and list) and returns the updated values (as if it is a functional-style API), which I think can be slightly confusing. I think it would be better to choose one style (functional, no side-effects) or the other (imperative, no return-value).

AllenTiTaiWang replied:
Done. I decided to go with no-return.

gramalingam commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,137 @@ def _export(
+
+
+@_beartype.beartype
+def _find_onnxscript_op_recursive(
```
Nit: is `recursive` needed in the name? That's just part of its implementation. I would just omit it. (I was also confused, because I thought `recursive` here meant that it handles transitively called functions, that is function called from functions called by this model).

AllenTiTaiWang replied:
Makes sense. Done.

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,137 @@ def _export(
+            if onnx_fn is not None:
+                # TODO(titaiwang): to_function_proto is onnx-script API and can be annotated
+                # after onnx-script is dependency
+                onnx_function_list.append(onnx_fn.to_function_proto())  # type: ignore[attr-defined]
```
Checking again, do we need this list deduplicated?

AllenTiTaiWang replied:
already did. That's what `included_node_func (set)` is doing. I think I had this comment from you somewhere above.

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,135 @@ def _export(
+        onnx_function_group = registration.registry.get_function_group(node_kind)
+        if (
+            node.domain
+            and not jit_utils.is_aten(node.domain)
```
nit: do we need to exclude all predefined namespaces (onnx, prim)? Changed to nit since I guess it's only here because of `aten` fallback.

AllenTiTaiWang replied:
Added is_onnx and is_prim to exclude corner case in registry (potential issue)

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -1660,6 +1625,135 @@ def _export(
+            if onnx_fn is not None:
+                # TODO(titaiwang): to_function_proto is onnx-script API and can be annotated
+                # after onnx-script is dependency
+                onnx_function_list.append(onnx_fn.to_function_proto())  # type: ignore[attr-defined]
```
nit: I think you can merge `onnx_function_list` and `included_node_func` by making the former a Set. Non-blocking ...

BowenBao reviewed (approved):
LGTM, some non blocking inline comments.

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1654,7 +1654,7 @@ def _export_file(
                 z.writestr(k, v)
     elif export_type == _exporter_states.ExportTypes.DIRECTORY:
-        if isinstance(io.BytesIO, f) or not os.path.isdir(f):  # type: ignore[arg-type]
+        if isinstance(io.BytesIO, type(f)) or not os.path.isdir(f):  # type: ignore[arg-type]
```
Is this right?

AllenTiTaiWang replied:
I had error message saying f doesn't fit the input requirement of isinstance()

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g

AllenTiTaiWang commented:
@pytorchbot rebase

titaiwangms committed:
Update on "[ONNX] Support custom Op with onnx-script local function"


Extend `register_custom_op` to support onnx-script local function. The FunctionProto from onnx-script is represented by custom op and inserted into ModelProto for op execution. 

NOTE: I did experiments on >2GB case of a simple model with large initializers:

```python
import torch

class Net(torch.nn.Module):
    def __init__(self, B, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm((B, C), eps=1e-3)
    def forward(self, x):
        return self.layer_norm(x)

N, B, C = 3, 25000, 25000
model = Net(B, C)
x = torch.randn(N, B, C)

torch.onnx.export(model, x, "large_model.onnx", opset_version=12)
```

And it turns out we won't get model_bytes > 2GB after `_export_onnx` pybind cpp function, as we split initializer in external files in that function, and have serialization before return the model bytes, which protobuf is not allowed to be larger than 2GB at any circumstances.


The test cases can be found in the next PR #86907 .


[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g
